### DISCUSSION TITLE
How does mitmweb "download" work? I can't see anything in browser network tab, including websocket

### DISCUSSION THREAD

**Discussion by marcosscriven** [mitmproxy/mitmproxy#6103] [Q&A] [answered]
I've just been using mitmproxy/web and very impressed. 

I'm trying to work out quite how the 'download' functionality works on a logged stream (particularly for a file).

I see in browser console a Tornado `update` websocket is setup, and messages about new streams go via that.

But when I click "download" on a stream for a file, the file magically downloads to my browser, but I see nothing in the network tab, or messages in the websocket. 

I'd also like to know if/where mitmweb stores the streams on the server side - is there a directory I can just get it to keep all files?

**Comment by mhils** [comment, marked as answer]
Flow contents are obtained via a regular HTTP request, there isn't a lot of magic going on: https://github.com/mitmproxy/mitmproxy/blob/91d40ebe26445318ed77a9e476efb55ab601bcbf/mitmproxy/tools/web/app.py#L485

> I'd also like to know if/where mitmweb stores the streams on the server side - is there a directory I can just get it to keep all files?

At the moment they are stored in memory. mitmdump is the only tool that doesn't grow in memory usage over time.

**Reply by marcosscriven** [reply]
Thanks @mhils - that makes a lot more sense. I was running `lsof` on the server side and scratching my head where these large files were being downloaded from!

I see now as well how the various handlers are registered in the Tornado framework, so easier to see what's going on.
